server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java

```diff
@@ -149,6 +151,10 @@ public void recoverToTarget(ActionListener<RecoveryResponse> listener) {
                 IOUtils.closeWhileHandlingException(releaseResources, () -> wrappedListener.onFailure(e));
             };
 
+            final boolean useRetentionLeases = shard.indexSettings().isSoftDeleteEnabled()
+                && shard.indexSettings().getIndexMetaData().getState() != IndexMetaData.State.CLOSE;
```

Comment:
We also need to deal with frozen indices here.

Reply:
The reason we guard against closed indices here is because replicated write actions are blocked on closed indices, so we cannot create or remove any retention leases and must weaken this invariant in that case:

https://github.com/elastic/elasticsearch/blob/d7f7ebc7e2e7202e819e75edb88dbb56b662dee7/server/src/main/java/org/elasticsearch/index/seqno/ReplicationTracker.java#L750-L765

However this does not apply to frozen indices, and this invariant holds for them. This is despite the fact that we can't replay any operations into a frozen index, but that fact is true today in `master` as well. Could you clarify what extra changes you think we need here for frozen indices?

Reply:
okay, thanks for the explanation.